packages/headless/doc-parser/src/controller-resolver.ts

```diff
@@ -48,35 +60,49 @@ function resolveControllerInitializer(entry: ApiEntryPoint, name: string) {
     );
   }
 
-  return resolveFunction(entry, fn, [0]);
+  return resolveFunction(entry, fn, [0], referencesCount);
 }
 
-function resolveUtility(entry: ApiEntryPoint, name: string) {
+function resolveUtility(
+  entry: ApiEntryPoint,
+  name: string,
+  referencesCount: InterfaceReferencesCount
+) {
   const fn = findApi(entry, name) as ApiFunction;
-  return resolveFunction(entry, fn, []);
+  return resolveFunction(entry, fn, [], referencesCount);
 }
 
 function isFunction(item: ApiItem): item is ApiFunction {
   return item.kind === ApiItemKind.Function;
 }
 
-function extractTypesFrom(initializer: FuncEntity, utils: FuncEntity[]) {
-  const instanceTypes = extractTypesFromInitializerInstance(initializer);
-  const utilTypes = extractTypes(utils).types;
+function extractTypesFrom(
+  initializer: FuncEntity,
+  utils: FuncEntity[],
+  referencesCount: InterfaceReferencesCount
+) {
+  const instanceTypes = extractTypesFromInitializerInstance(
+    initializer,
+    referencesCount
+  );
+  const utilTypes = extractTypes(utils, referencesCount).types;
```

Comment:
I seriously doubt that the extraction logic needs to be adjusted. The extraction is just looking for a certain object depth, and flattens/extracts an interface when it encounters the depth.

I think the changes needed to support recursive types can be contained inside the resolver code.